### DISCUSSION TITLE
SMS implementation

### DISCUSSION THREAD

**Discussion by milad-reyhaani** [neos-sdi/adfsmfa#349] [Q&A] [answered]
Hi,
I hope you are doing well.

How can I send the OTP with a SMS API?

Is there any free module for this or should I write this with a language program?

**Comment by milad-reyhaani** [comment]
I saw the source code.
<img width="333" alt="image" src="https://github.com/user-attachments/assets/21190771-13fc-44a0-a400-09f7310846ee">
In which source file should I insert my API? My SMS server is a HTTP URL that uses GET method and I should pass phone number and token key to this API.
Could you please explain how can I implement this?

**Comment by redhook62** [comment]
Hi @helloboy1980 

Yes, you can take inspiration from these examples.
The code that will be useful to you to develop your component is in the file Neos.IdentityServer.Multifactor.SMS.cs.
It all depends if you do One-Way or Two-Way. 
But in any case before sending the SMS you must generate a unique code, include this code in your message and return this code to the component so that it can do the validation (One-Way case)

**Comment by milad-reyhaani** [comment]
I think I should use One-Way. 
I am not familiar with the .NET codes.
This is my API endpoint that uses GET method .`https://example.com/api/sms/v1/send-single-sms?recipientPhoneNumber=%Phonenumber%&message=%TokenKey%`

Is it possible that you write and implement this requirement in Neos.IdentityServer.Multifactor.SMS.cs file and send final file to me?
I would appreciate it.

**Reply by redhook62** [reply]
I think you need to replace the different placeholders
%phonenumber% your secret code
%tokenkey% your service access key

**Comment by milad-reyhaani** [comment]
I think you did not understand my point.

I said if you could add this API endpoint in the Neos.IdentityServer.Multifactor.SMS.cs file because I am not familiar with .NET coding.

**Reply by redhook62** [reply]
Don't worry, I will send you a sample from a former client, probably tomorrow because I am on vacation.
However, send me your email to the address indicated in the source codes

**Reply by milad-reyhaani** [reply]
This is my email address: dumped !!

**Reply by redhook62** [reply]
Ok, I have your email address.
I'll send you the source code and the compiled version at the end of next week.
In the meantime I'll delete your email

**Reply by milad-reyhaani** [reply]
Thanks, I will wait for your response.

**Reply by redhook62** [reply]
I have sent you some questions to the email address you provided.
I would like you to provide me with the expected answers

**Reply by MilSN** [reply]
Dear @redhook62 
I hope this message finds you well. I am reaching out to kindly request your guidance on implementing SMS functionality for our project. We are currently exploring various approaches and would greatly appreciate your expertise and any recommendations you could share to help us proceed effectively.

If you could provide us with any documentation, best practices, or advice on tools and frameworks that would be suitable for our needs, it would be immensely helpful.

Thank you in advance for your support. I look forward to your response and any assistance you can provide.

Best regards,

**Reply by redhook62** [reply]
have sent you some questions to the email address you provided.
I would like you to provide me with the expected answers

**Reply by redhook62** [reply]
> Dear @redhook62 I hope this message finds you well. I am reaching out to kindly request your guidance on implementing SMS functionality for our project. We are currently exploring various approaches and would greatly appreciate your expertise and any recommendations you could share to help us proceed effectively.
> 
> If you could provide us with any documentation, best practices, or advice on tools and frameworks that would be suitable for our needs, it would be immensely helpful.
> 
> Thank you in advance for your support. I look forward to your response and any assistance you can provide.
> 
> Best regards,

You can take a look on the multiple samples provided with this project
regards

**Comment by aishamc** [comment]
If you need APIs that can be integrated easily, You can use MessageCentral Apis. 
Also there is free 1k otp.

**Comment by redhook62** [comment, marked as answer]
There are a good number of examples in the solution, 
It's up to you to activate them, and for this look at the powershell commands...

On the other hand, if you don't know how to code a solution, why not work with a professional service provider of your choice.
On my side it is obvious that I can't deliver implementations for each request and moreover for free. 
Working this way is my choice !
If you consider that the examples are bad, I then ask you to GitHub to have better ones.

regards
